pkg/receive/handler.go

```diff
@@ -397,10 +439,31 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 
 	tLogger := log.With(h.logger, "tenant", tenant)
 
+	if h.writeGate != nil {
+		tracing.DoInSpan(r.Context(), "receive_write_gate_ismyturn", func(ctx context.Context) {
+			err = h.writeGate.Start(r.Context())
+		})
+		if err != nil {
+			level.Error(tLogger).Log("err", err, "msg", "internal server error")
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		defer h.writeGate.Done()
+	}
+
 	// ioutil.ReadAll dynamically adjust the byte slice for read data, starting from 512B.
 	// Since this is receive hot path, grow upfront saving allocations and CPU time.
 	compressed := bytes.Buffer{}
 	if r.ContentLength >= 0 {
+		// If the content length is known, we can block the request based on
+		// max size limit here already and avoid growing the buffer.
+		sizeLimit := int64(h.options.WriteRequestSizeLimit)
```

Comment:
I wonder if we can isolate limiting logic into a dedicated function (apart from `h.writeGate`) and use that in `receiveHTTP`. Maybe to a function like `requestUnderLimit` which takes in request as argument and returns parsed request and bool.

As these are a lot of `ifs` which would happen every time a remote write request is received, I think benchmarking this logic would be nice! Separate `requestUnderLimit` function would allow this! 🙂

Reply:
Let me see what I can do to isolate the logic. 👍

I have no worries regarding the ifs, they should add such a tiny amount of time to the handler's execution that the benchmark would show mostly noise. Most of the time spent in limiting logic will be due to the metrics increment and writing the error response back. 

If others also worry about this I can try to work out on a benchmark and see whether it's useful.